Rework audio and haptic feedback of FlorisBoard

patrickgold opened the pull request:
This PR adds a completely revamped settings panel for audio and haptic feedback when pressing keys, doing gestures, etc. and fixes a lot of bugs / unexpected behavior etc.

### Audio/haptic feedback screen

The new audio/haptic screen is accessible where the old settings for sounds & vibration where located.

| Audio | Haptic |
|---------|----------|
| ![input-feedback-screen-audio](https://user-images.githubusercontent.com/19412843/128628456-26b5457c-6148-4d8a-8512-92fd09c2882b.jpeg) | ![input-feedback-screen-haptic](https://user-images.githubusercontent.com/19412843/128628464-787c15a6-f9ad-4787-916c-599806b2aec8.jpeg) |

The feature toggles may change and do not work fully yet, still working on it.

### Fixes

- Add option to have configurable audio/haptic feedback for repeated actions, such as holding down the delete key. Closes #1041
- Fix crash when either vibration duration or strength is set to `0`. Fixes #824 and fixes #859
- FlorisBoard's haptic/audio prefs can now work even if these are disabled at a system level, thus allowing for a more granular control of haptic feedback if one does not want to enable audio/haptics across all apps. Closes #217
- Additionally closes #652 and probably closes #1084

### Known issues currently
- When the phone is set to muted, even ignore system prefs can't help here and audio won't play at all. Fixed by adding a note that this is the cause.
- Some features don't respect feature toggles or are assigned to the incorrect ones. Fixed by implementing the feature toggles

---

| Old settings for reference |
|------|
| ![input-feedback-screen-old](https://user-images.githubusercontent.com/19412843/128628504-e8b412d7-885a-47a4-9674-f70a97142ec4.jpeg) |

Glitchy-Tozier commented:
Very nice!
This may be too off-topic for this PR, but what do you think about removing those long `(e.g…`-additions after the settings-titles? They can't be read in normal upright-orientation, so imo they aren't that helpful.

KLaurence commented:
I'm not a contributor, but I think the examples (the `e.g.`s) can probably be placed in the respective option's description.

patrickgold commented:
I agree, the e.g. stuff gets cut off, making it hard to read, moving it to the description would indeed be a solution.

patrickgold commented:
I've now moved the e.g.'s in the description, looks a lot better. Additionally I've fixed some bugs, adjusted default values and added input feedback to the Smartbar and the one-handed buttons (they were silent before). I've tested out the behavior and will now merge it into master, so I can have it in the next beta release for additional feedback and eventual bug reports on the new implementation.

KLaurence commented:
Were the types of vibration intended to be different in perceived strength? I find that, in my device, the other vibration types are weaker than the key press vibration.

Also, should further bug reports be reported in this pull request? If not, would there be a discussion thread?

patrickgold commented:
> Were the types of vibration intended to be different in perceived strength? I find that, in my device, the other vibration types are weaker than the key press vibration.

Yes, this is fully intended. I did this because in my testing it was very annoying that things such as the repeated action or moving swipe vibrate in full strength rapidly (it just feels so weird), so these vibrations are reduced in strength.

> Also, should further bug reports be reported in this pull request? If not, would there be a discussion thread?

That's a good question. A discussion thread will be made when I do the beta release, for now if you have a bug report just post it here.

patrickgold commented:
> Holding down the Delete button on the keyboard doesn't trigger the key repeated action vibration, even at full duration and strength (100ms, 100%).

Strange, it does for me. Does the preference  "Use vibrator directly" make any difference?

> The perceived strength doesn't vary in accordance with the "Vibration strength" slider value, no matter if its value is 1% or 100%. The only difference in behavior that I have found is that the keyboard doesn't vibrate at 0%.

What Android version are you on? Only Android 8.0 and newer support this value, before that version I can only specify the duration.

Glitchy-Tozier commented:
> Use vibrator directly

I'd like to point out that this setting, assuming this is its full name, could probably be phrased more expressively.

Continue on, gentleman!

Glitchy-Tozier commented:
I like that the current heading is short, but the one you suggested sounds better.
Does it get cut off?

Additionally, I don't think most people know what an API is 🌝.  Putting "Android" into the long description might make it easier to understand.

A non-tech-savy description might be sth. like "Trigger vibration directly instead of asking android to do it"

KLaurence commented:
> Does the preference "Use vibrator directly" make any difference?

It being on or off makes no difference, sadly.

> What Android version are you on?

My phone is on Android 10.

> "Control hardware vibration directly"
> What do you think of this title?

I think it's good. The alternative titles I've thought of are "Bypass system's vibration settings" or "Ignore system's vibration settings" but they aren't much shorter than your title.

Glitchy-Tozier commented:
> Ignore system's vibration settings

This sounds amazing. Is it correct, @patrickgold ?

patrickgold commented:
> It being on or off makes no difference, sadly.

Ok thanks for clarifying, will debug the input feedback manager tomorrow, maybe I can catch the bug.

> > Ignore system's vibration settings
>
> This sounds amazing. Is it correct, @patrickgold ?

Nope, it is not correct. There's a setting above the vibrator one which is named exactly like this and indeed checks a system setting if system-wide touch vibration is enabled or not. The use vibrator one literally decides if I call `<view>.performHapticFeedback()` or `vibrator.vibrate()`, the former one behaving like a b*tch on some devices while the latter one is much more stable, thus I give users the choice here.

**Another suggestion for the vibrator preference:**

Title: Trigger vibration directly
Summary: Trigger vibration directly instead of using Android's haptic feedback feature set

It has a short title but a non-tec-savy description IMO. What do both of you think of this option?

Glitchy-Tozier commented:
> What do both of you think of this option?

I like it a lot, go for it!  :)

KLaurence commented:
I like it. I also prefer it over "Control hardware vibration directly". The only thing I want to do is add why this option is necessary in the summary:

> **Trigger vibration directly**
> Trigger vibration directly instead of using Android's haptic feedback feature set that might not work properly on some devices

This summary could entice users to enable the option. I don't think there is a problem with enabling it unless it's unnecessary or even harmful, but I'm not sure if it is either of the two.

patrickgold commented:
> >  Does the preference "Use vibrator directly" make any difference?
> 
> It being on or off makes no difference, sadly.

I just discovered that while >=Android 8.0 devices supports providing an amplitude in the Vibrator service, many devices have no hardware amplitude control support and then the amplitude (which is basically the value of strength but adjusted) is just ignored. I will add this to the summary of the vibration strength preference.

patrickgold commented:
> I don't think there is a problem with enabling it unless it's unnecessary or even harmful, but I'm not sure if it is either of the two.

Enabling this is **never** harmful, as I just do a short one-shot vibration effect. Neither unnecessary, more like a personal preference of the user. The only devices where the haptic feedback API is actually good and even better than using the vibrator is on Google Pixel devices. Seems like Google devices have some well-implemented haptics on a hardware level, thus using the vibrator directly is a step back.

patrickgold commented:
#1150 is the "part 2" of this rework, which overhauls the mentioned UI strings and also adds some more descriptive summaries for the vibration strength and why it won't work (if your device is affected). @KLaurence as you have had this PR's debug artifact, could you also try out #1150's and test the settings panel of the audio/haptic panel? Thanks in advance!

KLaurence commented:
> I just discovered that while >=Android 8.0 devices supports providing an amplitude in the Vibrator service, many devices have no hardware amplitude control support and then the amplitude (which is basically the value of strength but adjusted) is just ignored. I will add this to the summary of the vibration strength preference.
>
> Enabling this is **never** harmful. Neither unnecessary, more like a personal preference of the user. 

These make sense.

> As you have had this PR's debug artifact, could you also try out #1150's and test the settings panel of the audio/haptic panel?

I'm currently having issues with downloading the artifact, but once the issues disappear, I'll update you.

KLaurence commented:
The vibration strength option is grayed out and says that my device doesn't support amplitude control, which explains the bug, frankly. The new "Trigger vibration directly" title and summary strings appear in my phone just fine too.

Other than that, nothing in the artifact's settings seems out of the blue for me.

Thank you for the bugfix!

patrickgold commented:
> The vibration strength option is grayed out and says that my device doesn't support amplitude control, which explains the bug, frankly. The new "Trigger vibration directly" title and summary strings appear in my phone just fine too.

I added this because else I suspect I will continue to get issues for the vibration strength and why it won't work.

Thanks for testing it out. I will merge the improvements in and they will be released in beta09 either tomorrow or on Sunday.
